dss/collections/__init__.py

```diff
@@ -0,0 +1,109 @@
+import os
+
+from dss.config import Replica
+from dss.util.aws.clients import dynamodb  # type: ignore
+
+
+_collections_db_table_template = f"dss-collections-db-{{}}-{os.environ['DSS_DEPLOYMENT_STAGE']}"
+
+
+class CollectionData:
+    REPLICA = 'replica'
+    OWNER = 'owner'
+    UUID = 'uuid'
+    VERSION = 'version'
+
+
+def put_collection(doc: dict):
+    """Adds a new owner associated collection to the table if it does not already exist."""
+    dynamodb.put_item(
+        TableName=_collections_db_table_template.format(doc[CollectionData.REPLICA]),
+        Item={
+            'hash_key': {
+                'S': doc[CollectionData.OWNER]
+            },
+            'sort_key': {
+                'S': doc[CollectionData.UUID]
+            },
+            'versions': {
+                'S': doc[CollectionData.VERSION]  # comma-delimited string
+            }
+        }
+    )
+
+
+def patch_collection(doc: dict):
+    """Edits an existing owner associated collection to append a new collection."""
+    current = get_collection(Replica[doc[CollectionData.REPLICA]], doc[CollectionData.OWNER], doc[CollectionData.UUID])
+    versions = ','.join([doc[CollectionData.VERSION]] + current[0]['collection_versions'])
+    dynamodb.update_item(
+        TableName=_collections_db_table_template.format(doc[CollectionData.REPLICA]),
+        Key={
+            'hash_key': {
+                'S': doc[CollectionData.OWNER]
+            },
+            'sort_key': {
+                'S': doc[CollectionData.UUID]
+            }
+        },
+        UpdateExpression='SET versions = :val1',
+        ExpressionAttributeValues={
+            ':val1': {
+                'S': versions  # comma-delimited string
+            }
+        }
+    )
+
+
+def get_collection(replica: Replica, owner: str, uuid: str):
+    """Returns all versions of a collection UUID for a user."""
+    db_resp = dynamodb.get_item(
+        TableName=_collections_db_table_template.format(replica.name),
+        Key={
+            'hash_key': {
+                'S': owner
+            },
+            'sort_key': {
+                'S': uuid
+            }
+        }
+    )
+    items = [db_resp.get('Item', None)] if db_resp.get('Item', None) else []
+    return collections_from_items(items)
+
+
+def collections_from_items(items: list) -> list:
+    collections = []
+    for item in items:
+        versions = [i.strip() for i in item['versions']['S'].split(',') if i.strip()]
+        collections.append({'collection_uuid': item['sort_key']['S'],
+                            'collection_versions': versions})
+    return collections
+
+
+def get_collections_for_owner(replica: Replica, owner: str) -> list:
```

Comment:
This method should support pagination.

See DynamoDB query pagination: https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Query.html#Query.Pagination

Reply:
Will do.